addons/sale/models/sale_order.py

```diff
@@ -1431,11 +1435,17 @@ def _get_invoiceable_lines(self, final=False):
                     # at the end of the invoice, in a specific dedicated section.
                     down_payment_line_ids.append(line.id)
                     continue
-                if pending_section:
-                    invoiceable_line_ids.append(pending_section.id)
-                    pending_section = None
+                if section_line_ids:
+                    if not line.display_type:
+                        has_invoicable_product = True
+                    section_line_ids.append(line.id)
+                    continue
```

Comment:
Why do you need the `has_invoice_product` ?
I believe that you can directly add the `section_line_ids` to the `invoiceable_line_ids` if you find a product line here :thinking: 

```suggestion
                if section_line_ids:
                    if line.display_type:
                        section_line_ids.append(line.id)
                    else:
                        invoiceable_line_ids.extend(section_line_ids)
                        section_line_ids = []
                    continue
```

Reply:
The `has_invoicable_product` flag was introduced to check for the presence of products in `section_line_ids` (a temporary list which holds all the lines in the current section). 
These lines are added to ` invoiceable_line_ids` only if the `has_invoicable_product` flag is set True, ensuring that the order of lines is preserved.

Reply:
Yes, but why can't you simply add all the lines in the temporary list and the product line once you find a product line?
Unless I'm missing something, what you do is that when you find a product, you set that variable to True, and keep adding the lines to the temporary list, until the end when you add the whole list to the invoiceable lines, and this gives the same result as adding the temporary list whenever you find the product, and empty it :thinking:

Reply:
Previously, I thought of working on section level but this seems more simplified.
Updated the code according to the suggested logic. Can you please have a look.
Thank you.